### PR TITLE
Add child component

### PR DESCRIPTION
**Pull request by AyeRlS**
Adding a dedicated component to add new content to a chapter or skill as asked in #217 

Please note that clicking the icons and the service behind it hasn't been implemented yet. Also, the little arrow above this component hasn't been added yet.

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]
The close button 'x' is at the place I say it was not appropriate to place it to 🤷‍♂️

File: src/app/modules/shared-components/components/add-content/add-content.component.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<alg-sub-section icon="sign-in-alt" label="Add a content" [isClosed]="state == 'closed'" (close)="clearInput()">
+  <alg-input [parentForm]="newContentForm" name="title" placeholder="Enter a title to create a new child"></alg-input>
+
+  <div *ngIf="state == 'opened'" class="new-content-type">
```
Comment (smadbe) [inline src/app/modules/shared-components/components/add-content/add-content.component.html:4]: All string comparison must use `===`

File: src/app/modules/shared-components/components/add-content/add-content.component.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<alg-sub-section icon="sign-in-alt" label="Add a content">
+  <alg-input [parentForm]="newContentForm" name="title" placeholder="Enter a title to create a new child" [hasClearButton]="true"></alg-input>
```
Comment (smadbe) [inline src/app/modules/shared-components/components/add-content/add-content.component.html:2]: the clear button should be shown only if something has been typed!

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.html [hunk 3]
```diff
@@ -42,13 +42,16 @@
           </div>
         </ng-template>
       </p-table>
+
+      <span *ngIf="!data.length">No children for this chapter.</span>
+
+      <alg-add-content [allowSkills]="this.itemData?.item?.type === 'Skill'"></alg-add-content>
```
Comment (smadbe) [inline src/app/modules/item/components/item-children-edit/item-children-edit.component.html:48]: * Don't use "this" in templates. 
* Item is not optional

File: src/app/modules/shared-components/components/input/input.component.scss [hunk 4]
```diff
@@ -63,9 +63,25 @@
...
       font-style: italic;
-      color: lightgray;
+      color: rgb(118, 118, 121);
       font-size: 1.2rem;
     }
+
+
+  }
+
+  .clear-button {
```
Comment (smadbe) [inline src/app/modules/shared-components/components/input/input.component.scss:73]: could you make it grey as the text and the left icon?

File: src/app/modules/shared-components/components/add-content/add-content.component.html (hunk 1 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/add-content/add-content.component.html]: Fixed

File: src/app/modules/shared-components/components/add-content/add-content.component.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  templateUrl: './add-content.component.html',
+  styleUrls: [ './add-content.component.scss' ],
+})
+export class AddContentComponent implements OnInit, OnDestroy {
+  @Input() allowSkills = false;
+
+  newContentForm: FormGroup = this.formBuilder.group({ title: '' });
+
+  state: 'opened' | 'closed' = 'closed';
+  subscription?: Subscription;
```
Comment (smadbe) [inline src/app/modules/shared-components/components/add-content/add-content.component.ts:16]: private

File: src/app/modules/shared-components/components/input/input.component.ts [hunk 6]
```diff
@@ -24,21 +23,28 @@ export class InputComponent implements OnInit, OnChanges {
   @Input() inputType = 'text';
   @Input() inputIcon = 'font'; // a font-awesome icon identifier
   @Input() buttonIcon?: string; // a font-awesome icon identifier for the input button
+  @Input() hasClearButton = false;
 
   @Output() click = new EventEmitter();
 
-  control: AbstractControl | null = null;
+  subscription?: Subscription;
```
Comment (smadbe) [inline src/app/modules/shared-components/components/input/input.component.ts:30]: private

File: src/app/modules/shared-components/components/input/input.component.ts [hunk 7]
```diff
@@ -24,21 +23,28 @@ export class InputComponent implements OnInit, OnChanges {
...
+  value = '';
```
Comment (smadbe) [inline src/app/modules/shared-components/components/input/input.component.ts:31]: what do you use this value for?

File: src/app/modules/shared-components/components/input/input.component.html [hunk 8]
```diff
@@ -5,6 +5,11 @@
     [ngClass]="{'has-button': buttonIcon}"
     [formControlName]="name"
   >
+  <span *ngIf="hasClearButton" class="clear-button">
+    <button (click)="clearInput()">
```
Comment (smadbe) [inline src/app/modules/shared-components/components/input/input.component.html:9]: Couldn't you just put a template variable on the input and do `#input.value=''` here?

**Review by smadbe — CHANGES_REQUESTED** [review]

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.html (hunk 3 above)
Comment (AyeRlS) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.html:48]: image
I really understand why but it won't let me compile without that '?'

Comment (AyeRlS) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.html:48]: My bad, I forgot that '?'  doesn't work in the same way it does in a ts file. Can I use '!' here?

File: src/app/modules/shared-components/components/input/input.component.html (hunk 8 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/input/input.component.html:10]: Done

File: src/app/modules/shared-components/components/input/input.component.ts (hunk 7 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/input/input.component.ts]: Use to verify that the length of the input and show the 'x' if there is something. But I've removed it to use a template variable

File: src/app/modules/shared-components/components/input/input.component.ts (hunk 6 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/input/input.component.ts]: It has been removed

File: src/app/modules/shared-components/components/add-content/add-content.component.ts (hunk 5 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/add-content/add-content.component.ts]: Done

File: src/app/modules/shared-components/components/input/input.component.scss (hunk 4 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/input/input.component.scss:73]: Done

File: src/app/modules/shared-components/components/add-content/add-content.component.html (hunk 2 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/add-content/add-content.component.html:2]: It was supposed to be already done. Fixed

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.html (hunk 3 above)
Comment (AyeRlS) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.html]: Using a '!' can cause problem if this component is used incorrectly so I've used another way

File: src/app/modules/shared-components/components/input/input.component.html (hunk 8 above)
Comment (smadbe) [reply src/app/modules/shared-components/components/input/input.component.html:10]: My remark was for `(click)="clearInput()"`

**Comment by smadbe** [comment]
Bug: I type 3 char, I press "x", the chapter/task button are still displayed

File: src/app/modules/shared-components/components/input/input.component.html (hunk 8 above)
Comment (AyeRlS) [reply src/app/modules/shared-components/components/input/input.component.html:10]: After testing, the `#input.value=''` doesn't feat our needs, I've reverted to what was done previously.

As it is not the cleanest way to do it, in the future we should look into alternatives to see if there is a better way to do things, like using [this](https://angular.io/api/forms/DefaultValueAccessor)

**Review by smadbe — APPROVED** [review]

**Pull request merged**